vulture_os/toolkit/api_parser/ubika/ubika.py

```diff
@@ -0,0 +1,279 @@
+                    logs.extend(newLogs)
+                last_save_token = next_token
+            else:
+                page_end = True
+
+        return logs
+
+    def format_log(self, log: dict) -> str:
+        request_headers : list = log.get('request', {}).get('headers', [])
+        new_request_headers = {}
+        if request_headers:
+            for header in request_headers:
+                new_key = header.get('key').lower()
+                new_value = header.get('value')
+                if new_key and new_value:
+                    new_request_headers[new_key] = new_value
+        log['request']['headers'] = new_request_headers
+
+        request_cookies : list = log.get('request', {}).get('cookies', [])
+        new_request_cookies = {}
+        if request_cookies:
+            for cookie in request_cookies:
+                new_key = cookie.get('key').lower()
+                new_value = cookie.get('value')
+                if new_key and new_value:
+                    new_request_cookies[new_key] = new_value
+        log['request']['cookies'] = new_request_cookies
+
+        context_reaction : str = log.get('context', {}).get('reaction')
+        if context_reaction:
+            log['context']['reaction'] = context_reaction.lower()
+
+        return dumps(log)
+
+    def test(self):
+        logs = []
+
+        try:
+            if (not self.ubika_access_token) or (self.ubika_access_expires_at <= (timezone.now() + timedelta(minutes=1)).timestamp()):
+                self.login()
+
+            namespace = self.ubika_namespaces[0]
+            for kind in ["security", "traffic"]:
+                if kind == "security":
+                    url = self.security_events_endpoint
+                elif kind == "traffic":
+                    url = self.traffic_logs_endpoint
+                url = url.replace("__NAMESPACE__", namespace)
+
+                logger.info(f"[{__parser__}][test]:Running tests...", extra={'frontend': str(self.frontend)})
+
+                since = timezone.now() - timedelta(minutes=15)
+                to = timezone.now() - timedelta(minutes=5)
+
+                newLogs = self.get_logs(url, since, to)
+                logs.extend(newLogs)
+
+                msg = f"{len(logs)} logs retrieved"
+                logger.info(f"[{__parser__}][test]: {msg}", extra={'frontend': str(self.frontend)})
+
+                return {
+                    "status": True,
+                    "data": logs[0:99]
+                }
+        except Exception as e:
+            logger.exception(f"[{__parser__}][test]: {e}", extra={'frontend': str(self.frontend)})
+            return {
+                "status": False,
+                "error": str(e)
+            }
+
+    def execute(self):
+        for namespace in self.ubika_namespaces:
+            if (not self.ubika_access_token) or (self.ubika_access_expires_at <= (timezone.now() + timedelta(minutes=1)).timestamp()):
+                self.login()
```

Comment:
may move the condition in `login()` and the call to `login()` in `__execute_query()`